File: medium/739-daily-temperatures.ts
function dailyTemperatures(temperatures: number[]): number[] {
  // const results = Array(temperatures.length).fill(0);
  // const stack: number[] = [];
  // for(let i = 0; i < temperatures.length; i++) {
  //   while(stack.length && temperatures[i] > temperatures[stack[stack.length - 1]]) {
  //     const temp = stack.pop();
  //     if(temp !== undefined) results[temp] = i - temp;
  //   }
  //   stack.push(i);
  // }
  // return results;

  const stack = [], ans = new Array(temperatures.length).fill(0);
  for(let i = 0; i < temperatures.length; i++){
    while(stack.length && temperatures[stack[stack.length-1]] < temperatures[i]){
      ans[stack[stack.length-1]] = i - stack[stack.length-1];
      stack.pop();
    }
    stack.push(i);
  }
  return ans;
};

console.log(dailyTemperatures([73,74,75,71,69,72,76,73]))

export default dailyTemperatures;

File: python/739-daily-temperatures.py
from typing import List


def dailyTemperatures(temperatures: List[int]) -> List[int]:
    stack = []
    ans = [0] * len(temperatures)
    # for i in range(len(temperatures)):
    #     while stack and temperatures[stack[-1]] < temperatures[i]:
    #         ans[stack[-1]] = i - stack[-1]
    #         stack.pop()
    #     stack.append(i)

    for i, temperature in enumerate(temperatures):
        while stack and temperatures[stack[-1]] < temperature:
            index = stack.pop()
            ans[index] = i - index
        stack.append(i)
    return ans


print(dailyTemperatures([73, 74, 75, 71, 69, 72, 76, 73]))
